src/phpDocumentor/Dsn.php

```diff
@@ -0,0 +1,232 @@
+
+    /**
+     * Returns the query part of the DSN
+     *
+     * @return string[]
+     */
+    public function getQuery()
+    {
+        return $this->query;
+    }
+
+    /**
+     * Returns the parameters part of the DSN
+     *
+     * @return string[]
+     */
+    public function getParameters()
+    {
+        return $this->parameters;
+    }
+
+    /**
+     * Parses the given DSN
+     *
+     * @param string $dsn
+     */
+    private function parse($dsn)
+    {
+        if (! is_string($dsn)) {
+            throw new \InvalidArgumentException(
+                sprintf('"%s" is not a valid DSN.', $dsn)
+            );
+        }
+
+        $dsnParts = explode(';', $dsn);
+        $location = $dsnParts[0];
+        unset($dsnParts[0]);
+        $locationParts = parse_url($location);
+
+        if (! isset($locationParts['scheme'])) {
+            $locationParts['scheme'] = 'file';
+            $location = 'file://' . $location;
+        }
+
+        if (! filter_var($location, FILTER_VALIDATE_URL)) {
+            throw new \InvalidArgumentException(
+                sprintf('"%s" is not a valid DSN.', $dsn)
+            );
+        }
+
+        $this->parseScheme($locationParts);
+
+        $this->parseHostAndPath($locationParts);
+
+        $this->port = isset($locationParts['port']) ? $locationParts['port'] : "";
+
+        $this->user = isset($locationParts['user']) ? $locationParts['user'] : "";
+
+        $this->password = isset($locationParts['pass']) ? $locationParts['pass'] : "";
+
+        $this->query = isset($locationParts['query']) ? explode('?', $locationParts['query']) : "";
+
+        $this->parseParameters($dsnParts);
+    }
+
+    /**
+     * validates and sets the scheme property
+     *
+     * @return void
+     */
+    private function parseScheme($locationParts)
+    {
+        $validSchemes = ['file', 'git+http', 'git+https'];
+        if (! in_array(strtolower($locationParts['scheme']), $validSchemes)) {
+            throw new \InvalidArgumentException(
+                sprintf('"%s" is not a valid scheme.', $locationParts['scheme'])
+            );
+        }
+        $this->scheme = strtolower($locationParts['scheme']);
+    }
+
+    /**
+     * Validates and sets the host and path properties
+     *
+     * @return void
+     */
+    private function parseHostAndPath($locationParts)
+    {
+        $path = isset($locationParts['path']) ? $locationParts['path'] : "";
+        $host = isset($locationParts['host']) ? $locationParts['host'] : "";
+
+        if ($this->getScheme() === 'file') {
+            $this->path = $host . $path;
+        } else {
+            $this->host = $host;
+            $this->path = $path;
+        }
+    }
+
+    /**
+     * validates and sets the parameters property
+     *
+     * @return void
+     */
+    private function parseParameters($dsnParts)
+    {
+        if (!empty($dsnParts)) {
+            foreach ($dsnParts as $part) {
+                if (! strpos($part, '=')) {
```

Comment:
small performance tip: you can do the explode before this check and instead of checking using strpos you can count the $option array; if it isn't 2 then the part is not valid